Revocation on deletion

adithyaphilip opened the pull request:
Sorry for the delay, took me a while to get the workflow right.
* Making the dialog appear due to a pendingResult in DeleteOperation made the code in it too unwieldy. Instead, revocation now happens in a new `RevokeOperation`, and `DeleteKeyDialogActivity` determines what dialog to show.
* Added atomic/non-atomic upload to `EditKeyOperation`
* `mCachePassphrase` in `CryptoInputParcel` for operations to specify if they want their passphrase cached.
* Allowed operations to modify the `CryptoInputParcel` they receive - this was primarily so that `EditKeyOperation` could save the modified keyring (so it wouldn't have to repeat the operation) in case the uploading returned a pendingResult. However, this required making `UncachedKeyRing` and related `spongycastle` classes Serializable/Parcelable, so it isn't implemented as of now. Currently, this feature is only used by `RevokeOperation` to specify that it does not want its passphrase to be cached.
* Added retry on upload failure as a type of pendingResult, for operations that require atomic upload. Downside is an operation which uses this won't show an error log if the user decides not to retry, hence the wording "Cancel Operation" instead of just "Cancel" to make it clear. One workaround I'm thinking of going ahead with is adding an `OperationResult` to the `CryptoInputParcel` that is a part of the pendingResult. This could be shown in case of cancellation.

Screenshots:
![screenshot_2015-07-10-08-10-40](https://cloud.githubusercontent.com/assets/5712241/8611412/32598832-26de-11e5-9001-33923e77d05f.png)
![screenshot_2015-07-10-06-57-56](https://cloud.githubusercontent.com/assets/5712241/8610656/a96b5678-26d2-11e5-8cac-df10a2386707.png)

dschuermann commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/results/InputPendingResult.java:
```diff
@@ -30,26 +29,39 @@
     }
 
-    public InputPendingResult(OperationLog log, RequiredInputParcel requiredInput) {
+    /* public InputPendingResult(OperationLog log, RequiredInputParcel requiredInput) {
```
Looks like this is some leftover code which can be removed

dschuermann commented:
First: The UI looks great!

I tried with a key that was until now only locally and not uploaded to a keyserver. It showed a success for revoke and upload but there were some serious problems:
```
 22174             Keychain D  D  Everything okay!
 22174             Keychain D  D  Not caching entered passphrase!
 22174         OpenGLRenderer  D  endAllStagingAnimators on 0x9d351300 (RippleDrawable) with handle 0x9d982d60
 22174             Keychain D  D  received activity result in OperationHelper
 22174             Keychain D  D  received activity result in OperationHelper
 22174             Keychain D  D  handling activity result in OperationHelper
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_REVOKE, mParameters=[749c 89b8 0f6d e2e4], mIndent=0}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, proj=[has_secret])
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/unified, proj=[fingerprint])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_ED, mParameters=null, mIndent=0}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_ED_FETCHING, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/unified, proj=[has_any_secret, verified, privkey_data])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_MF, mParameters=[0x749c89b80f6de2e4], mIndent=0}
 22174             Keychain D  D  Send message by setProgress with progress=10, max=100
 22174             Keychain D  D  Send message by setProgress with progress=10, max=100
 22174             Keychain D  D  Send message by setProgress with progress=15, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_MF_UNLOCK, mParameters=null, mIndent=1}
  741           WindowManager  V  Adding window Window{3bd4e264 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.DeleteKeyDialogActivity} at 24 of 32 (before Window{36212fa0 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity EXITING})
 22174             Keychain D  D  Send message by setProgress with progress=37, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=INFO, mType=MSG_MF_SUBKEY_REVOKE, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174               Timeline  I  Timeline: Activity_idle id: android.os.BinderProxy@26c54161 time:544857333
 22174             Keychain D  D  Send message by setProgress with progress=60, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_MF_SUCCESS, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_MF, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_MF, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  D  Send message by setProgress with progress=0, max=1
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_EXPORT_UPLOAD_PUBLIC, mParameters=[0x749c89b80f6de2e4], mIndent=0}
  283          SurfaceFlinger  W  couldn't log to binary event log: overflow.
 22174             Keychain D  D  hkp keyserver add: https://hkps.pool.sks-keyservers.net:443/pks/add
 22174             Keychain D  D  params: keytext=-----BEGIN+PGP+PRIVATE+KEY+BLOCK-----%0A%0AlQdGBFWcP4oBEACpDJw38kYB%2F5d706DdloqTbJlseg%2B7OLI4VA0WGwIV6fdL0ryQ%0ApOZkL9Dpn9uY3hnXOvul7B%2Fs5GAsG45YN9KzT9Wcvf1fcw0zZxO2CfamGMMg6Il0%0Ad4RhZmx6FLdfdp14EdR5L0cLPHxwAHI3%2BuXa7tdWWzxajAuIN5FvMAxmqUHegTiH%0ABJPYWLtKoE%2FdVlEpug00GzHv3myltI2tw59i7wm6opO2zifF1SVRX4yBR6FrNdpI%0AW0PBIkz59Zr56VvJDUvFfqcF5JGISo6pMR7Lci0kjsOnyMdZy2XX3s63kzG0KCM0%0Avl2oFUBlu3T8%2BfuEryuxCzDKcDj2CFf%2FbrGF0IO53R6BupN6U9st9x4ONQaDlZ5N%0A61b8lL%2FywiLJw0t6ZZUtVmLZgc2qMcyA15S2Y8%2FvR132hOgmWVhq97%2Fngq4v2uBy%0A8cdIcc5aKrrbTCFw4t7ziZqbUUQ3MVtqh9dbgIkCfdOZbY%2BD3Gfek8Q9U4wjirrJ%0A0uLCHuFqFEI%2FZalR5EzNKjtyK%2BQTYLbNah6zFro0cmrNpFwg%2FYVviJxyJe2I1Pa2%0A%2FYcln35xufwZ2AsPoTKJqplCd2xNf6O85%2BSv%2FXwNT2u8pVvd3bcnToxPJ8lrU38n%0AZ3jzFYWsE7AAqy5xTnLMNUkXEx0hcWq%2BU026qBHaed4tjS%2F4Ij2S2MaeXQARAQAB%0A%2FgkDCC%2FqGh%2BVXXs1kNH1%2BTceC1siFQaeKd6bXyswqyOrF13k32rZoXUxAzl0n1m1%0ADjkVehuodT1HMsO8B32Glx5SREQ4CNBKXv%2F4FiS08Ek3ov7VhYCvNXnDulHJRbA4%0A9qo1K7dZjJn%2BlQm%2BBK92XRx2E2jBO5i84P2QrefBhlxUcQRgIDlMWC2ZOATphaXF%0AlQgROPRhfazKjNUiDvuTMcVGW%2FdgsdHliyq1P7MfsH0pLRWmSyZHc3ne3OxZIX4r%0A4iFD%2BVSuxdARc9CzLqR%2BjPfXRw34GWrUwpl9KyGYPpzXNhVun3uDCDlpUxzgUDs%2F%0A2mraB3uBXzoi%2B748a4p5Pvj67OfJKmDY58LrYeVwNs0wW4z6kKR4jQyt4bmcF9zQ%0AAyRzqjimHaK8RviweSDAbzU5oQaS6Yw4cSM3T7H8%2FTrrOHpoj93rKGFwsprPlNP%2F%0AXcVVYo9VuU3hXs%2B%2BA9%2F9m0NvA3jfLpztLSLg4NPk1DeiAfNLJFQDFdPXSqqxe4O3%0A18W7GYObp%2F4EclM1oD7FgbSRaB%2BweUD51JDMPTC9wei3HErN1zJB4P1kijxoxQo2%0Axvxvo6202YvCcRZFXYhLdDRWPK0nORLkMDz9x56XuOUC2bgT6TIPv0mOjE%2Bvz4%2Fh%0AZbZMC6LCnSp3jZLSXbZe%2BiUG2l1W%2FOQfqqJB364osx4xNtoAt0eTLpeOgeN3z4yB%0Aw10RG0JsGrx4nJhhaQpfPY5OZ3NDkkPpNTEvIvjMC6BKlDNbo0Hdp%2FbD7Ugo%2Bwq4%0A1bVaMOtZcmHzSqOgR%2B4Xogh%2B9VaO9H%2FqeS5J59exKsm9e9E%2B%2Fwqrg72IuDJtc6Pm%0Aigta5G5cEbExA%2BY9RFqvLUEgn48NGlCDAu4fctb0ewiKGWUjoeq4ilEfwoCfB1re%0A1W8x9326Dddgm%2FO%2BYf%2BzmNmIGR83fM1fTuurjgySvxLlcJlLJ0BiIhAEV5Bzf%2Ban%0A8o%2BeipnI1ip6eHSIXSbg1oYjZgviEn6bd4RboNisRHekbme2K%2FbRyFPW6cL20s3U%0AskhQ8%2BXhMdreA6dSADc6nhESVYwBnD7j59SaFTpsel4XFc67yH46oKcJ8iHyPuf%2B%0AGZ%2BhrW0o5rV3H9303bCoQLcbpY4%2F6Z9k7flDpMWp%2B1Z9y4fucT8IU9vJAkbLhdI9%0A25ZNmLgGle02pHNHfZS9p5hJ3j%2FPzRPcRICR%2B2Aue3gO5NqpOz4tbOYQCZkgomG8%0APBCGBNZPf4jTLtwOQm4sgEuxQCsIrcaYraRXl%2Fgl6kbPobiFP385Gt%2FFI%2FCJDrAO%0A2%2Ff36hHrRqup2C8eA9cBgbhSTV7EKMNnCRuK9Kb8Cw1Mvto5JoncwyVU86sA7zBz%0AWsHCGVH%2BmBkmbJ9S38gzoaFaBJaL27aAW6GKfyfflDZpIPRm7Z54m92h%2FXnW%2BsUR%0AJsFxBrd4kzzfcgQyBj4MdOm%2BV%2FeQ12Ju6SJx42CU1HBdxTMzEMueLU%2Fw63lnXzX1%0AKHfkTwtJweZYFTk5CO3fotmQeoA47E3NiOo%2FdjoTyRNpFW5bj6AFSvA9jI45y0%2FS%0AUa1tJD4GgaDUH71hC%2Fl654wTnRmIc5v8QSRlxWWI47Apalk3JH1MjMQD6MvUfRm%2B%0AybfdCV87FV%2BRvKg4j%2Bv37Us2tpmTB31dshYbGqhXBLNSTcStdWCHDqxI%2B9ogx7VJ%0AdAPlk0tE9QySyQYOuaa8cgQ0C8aCqSY%2FtwgzmXzKRBlmWh2pg1W18cIfO3rJqrBD%0ATYhtsuJZDb8BaJxVvk4Udj3Nd0UEWuLrsboF9SlAZNojHe0ExhUbFRiJAhwEIAEI%0AAAYFglWfoTcACgkQdJyJuA9t4uRQSw%2F%2FUh4lP6dId3QFIKlSb7PoEP8fwj7KWBma%0AGCe4bGOszxDhAYKy0VwZ1mAsB0oE6d8AMMJzh6lAWP6qkj2b%2FcuIOHu6gLnEaOIn%0AL7SQb%2FfdgCd3ptOOkEZOOjzKL0OsXDNWPZj0hKn5x9sL2mmtVSYF09BOI9EtXtVA%0AtwR31blir3hcU4Xf9bCAUTqt1AEdiEFZOW4ERtvoU%2Bntv4gtVZEtUypf0F9v3Ddh%0A7gsP1rNTQj5kpzNkc9RKJzRy686h9LGM9e3f4H2Nwp8OK4ks9yMWZBSljqmSVlSv%0Aej1gXdpkrrB9RqcXq%2BEeXGSocAUCVWTRy38VWm9drys31ykzIOLs21MPQDWArL%2FZ%0AeTyGiJTLtgqpexIWNtB5oyzDPG3kDuk6AeTb31zk6rJXiv2IbDp1xyzvrxFiAMKT%0AgrZ3VhKMBy47ap0qzGNEKylUUbVOWWFvd6je3iLFJxlxX9TFh%2F3nwrNaqLDamdAk%0A%2BM81Hs7UgpscCzLWF0aPYFHKuSdQgZHT%2BBDSuP8djIR%2Fb%2BJIb0eZbBceJBdGPao9%0AD09xWQTgsHPEJYLxvCRE%2B1u%2FSEw2DEUoIRyA8TTKHFZLBSEDSKgJ%2Bc%2BYGFRWDDs1%0Ah6iqtqZA1VGHnflKk0kP8xDhKWzBgFVaPGK4V6j4dnGoJa1n5JINkwTsThLqj8wT%0AA0awzdHG97C0G0FsaWNlIDIgPGFsaWNlQGV4YW1wbGUuY29tPokCOAQTAQgAIgUL%0ACQgHCgcVCAoJCwIDBBYBAgMCGQEFglWcP4oCngECmwEACgkQdJyJuA9t4uTgJxAA%0AnqNL0wcN4Ns3uZfQ3k8JLDVHK3TMBT6jjiFPSqiObkxd2SbSgjQML3yGELQLCms0%0A2kmzyTaVADt4tuu1WxNh4LoUkTQ6EAOtbLIwjovqtZYyS0c54xeUhCG5DrXXWSA4%0AeUbCPAGjNZvoQ6mU%2FUeO9gLQBgfi1NwonNzNVb1KssF8XHTJ12Lb%2FPhQbHK9l2cz%0AgEBekbIGhFR3fCVq159DqQWvjfsyl%2BORBbH3jhiaIwOdOwKmarS1oroxVkuzOQng%0A626JFRsz4%2Fq8ar17MYadWk9LOFq9%2FEwFvmd1h77SplbvRdCj
  741                Timeline  I  Timeline: Activity_windows_visible id: ActivityRecord{13d1d3a5 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.DeleteKeyDialogActivity t2752} time:544857496
  741     ConnectivityService  D  updateNetworkScore for NetworkAgentInfo [WIFI () - 472] to 60
  1159   ConnectivityManage..  D  CM callback handler got msg 524290
  741     ConnectivityService  D  rematching NetworkAgentInfo [WIFI () - 472]
  741     ConnectivityService  D  Network NetworkAgentInfo [WIFI () - 472] was already satisfying request 1. No change.
  741     ConnectivityService  D  notifyType AVAILABLE for NetworkAgentInfo [WIFI () - 472]
 12515            bs_taimport  W  Failed to read unit 2561: Unit not found
 12515            bs_taimport  W  Failed to read unit 2562: Unit not found
 12515            bs_taimport  W  Failed to read unit 2563: Unit not found
 12515            bs_taimport  W  Failed to read unit 2760: Unit not found
 12515            bs_taimport  W  Failed to read unit 2761: Unit not found
 12515            bs_taimport  W  Failed to read unit 2762: Unit not found
 12515            bs_taimport  W  Failed to read unit 2763: Unit not found
 12515            bs_taimport  W  Failed to read unit 2750: Unit not found
 12515            bs_taimport  W  Failed to read unit 2752: Unit not found
 12515            bs_taimport  W  Failed to read unit 2753: Unit not found
 12515            bs_taimport  W  Failed to read unit 2756: Unit not found
 12515            bs_taimport  W  Failed to read unit 2758: Unit not found
 22174             Keychain D  D  response code: 500
 22174             Keychain D  D  answer: <?xml version="1.0" encoding="utf-8"?>
 22174             Keychain D  D  <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd" >
 22174             Keychain D  D  <html xmlns="http://www.w3.org/1999/xhtml">
 22174             Keychain D  D  <head>
 22174             Keychain D  D  <title>Error handling request</title>
 22174             Keychain D  D  <meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
 22174             Keychain D  D  <style type="text/css">
 22174             Keychain D  D  /*<![CDATA[*/
 22174             Keychain D  D  .uid { color: green; text-decoration: underline; }
 22174             Keychain D  D  .warn { color: red; font-weight: bold; }
 22174             Keychain D  D  /*]]>*/
 22174             Keychain D  D  </style></head><body><h1>Error handling request</h1>Error handling request. Exception raised.</body></html>
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_EXPORT_UPLOAD_SUCCESS, mParameters=null, mIndent=1}
 22174             Keychain D  D  Send message by setProgress with progress=1, max=1
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_EXPORT_UPLOAD_PUBLIC, mParameters=[0x749c89b80f6de2e4], mIndent=2}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_EXPORT_UPLOAD_PUBLIC, mParameters=[0x749c89b80f6de2e4], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_IS, mParameters=[0x749c89b80f6de2e4], mIndent=0}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/unified, proj=[has_any_secret, verified, privkey_data])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IS_MERGE_SECRET, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_MG_UNCHANGED, mParameters=null, mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_KC_SECRET, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_MASTER, mParameters=[0x749c89b80f6de2e4], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_SUB, mParameters=[0x8c700d644a3867c2], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_SUB, mParameters=[0x5b86b08ef7073d77], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_KC_SUCCESS, mParameters=null, mIndent=2}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/unified, proj=[has_any_secret, verified, pubkey_data])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IS_MERGE_PUBLIC, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_MG_FOUND_NEW, mParameters=[1], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_KC_PUBLIC, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_MASTER, mParameters=[0x749c89b80f6de2e4], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_SUB, mParameters=[0x8c700d644a3867c2], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_KC_SUB, mParameters=[0x5b86b08ef7073d77], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_KC_SUCCESS, mParameters=null, mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_PREPARE, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_INSERT_KEYRING, mParameters=null, mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_INSERT_SUBKEYS, mParameters=null, mIndent=2}
 22174             Keychain D  D  Send message by setProgress with progress=74, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_MASTER, mParameters=[0x749c89b80f6de2e4], mIndent=3}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_MASTER_FLAGS_CXXX, mParameters=null, mIndent=4}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_SUBKEY, mParameters=[0x8c700d644a3867c2], mIndent=3}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_SUBKEY_FLAGS_XXSX, mParameters=null, mIndent=4}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_SUBKEY, mParameters=[0x5b86b08ef7073d77], mIndent=3}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_SUBKEY_FLAGS_XEXX, mParameters=null, mIndent=4}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/unified, proj=[master_key_id, has_any_secret, verified, pubkey_data])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_UID_CLASSIFYING, mParameters=[10], mIndent=2}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_UID_PROCESSING, mParameters=[Alice 2 <alice@example.com>], mIndent=3}
 22174             Keychain D  D  Send message by setProgress with progress=82, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_UID_REORDER, mParameters=null, mIndent=2}
 22174             Keychain D  V  delete(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/public)
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/unified, proj=[_id, master_key_id, user_id, is_revoked, is_expired, verified, has_any_secret, has_duplicate_user_id, fingerprint, creation])
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_DELETE_OLD_OK, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IP_APPLY_BATCH, mParameters=null, mIndent=1}
 22174             Keychain D  D  Send message by setProgress with progress=86, max=100
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/public, values=key_ring_data=[B@22bfaa91 master_key_id=8402742428392350436)
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=can_certify=true is_revoked=true can_authenticate=false master_key_id=8402742428392350436 fingerprint=[B@2eb9fdf6 rank=0 can_sign=false key_curve_oid=null can_encrypt=false key_id=8402742428392350436 key_size=4096 algorithm=1 creation=1436303242)
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=can_certify=false is_revoked=false can_authenticate=false master_key_id=8402742428392350436 fingerprint=[B@2973c64 rank=1 can_sign=true key_curve_oid=null can_encrypt=false key_id=-8327140986614945854 key_size=4096 algorithm=1 creation=1436303242)
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=can_certify=false is_revoked=false can_authenticate=false master_key_id=8402742428392350436 fingerprint=[B@1d800382 rank=2 can_sign=false key_curve_oid=null can_encrypt=true key_id=6595152832406961527 key_size=4096 algorithm=1 creation=1436303242)
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/user_ids, values=is_revoked=false attribute_data=null is_primary=true master_key_id=8402742428392350436 rank=0 type=null user_id=Alice 2 <alice@example.com>)
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/certs, values=key_id_certifier=8402742428392350436 master_key_id=8402742428392350436 verified=1 rank=0 data=[B@91f4a93 type=19 creation=1436303242)
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_IP_SUCCESS, mParameters=null, mIndent=1}
 22174             Keychain D  D  Send message by setProgress with progress=91, max=100
 22174             Keychain D  D  Send message by setProgress with progress=91, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_IS, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  D  insert(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/secret, values=key_ring_data=[B@2bbdbad0 master_key_id=8402742428392350436)
 22174             Keychain D  V  update(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=has_secret=1)
 22174             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_IS_IMPORTING_SUBKEYS, mParameters=null, mIndent=2}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/unified, proj=[_id, master_key_id, user_id, is_revoked, is_expired, verified, has_any_secret, has_duplicate_user_id, fingerprint, creation])
 22174             Keychain D  V  update(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=has_secret=2)
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_IS_SUBKEY_OK, mParameters=[0x749c89b80f6de2e4], mIndent=3}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/unified, proj=[_id, master_key_id, user_id, is_revoked, is_expired, verified, has_any_secret, has_duplicate_user_id, fingerprint, creation])
 22174             Keychain D  V  update(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=has_secret=2)
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_IS_SUBKEY_OK, mParameters=[0x8c700d644a3867c2], mIndent=3}
 22174             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/unified, proj=[_id, master_key_id, user_id, is_revoked, is_expired, verified, has_any_secret, has_duplicate_user_id, fingerprint, creation])
 22174             Keychain D  V  update(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/8402742428392350436/keys, values=has_secret=2)
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_IS_SUBKEY_OK, mParameters=[0x5b86b08ef7073d77], mIndent=3}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_IS_SUCCESS, mParameters=null, mIndent=2}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_IS, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_IS, mParameters=[0x749c89b80f6de2e4], mIndent=1}
 22174             Keychain D  D  Send message by setProgress with progress=100, max=100
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_ED_SUCCESS, mParameters=null, mIndent=0}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_ED, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: SubLogEntryParcel{mLevel=START, mType=MSG_ED, mParameters=null, mIndent=1}
 22174             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_REVOKE_OK, mParameters=null, mIndent=1}
```

Looks like it's uploading the private key, not the public key ;) and it wasn't really handling the return code 500.

dschuermann commented:
Regarding the revocation procedure:
Instead of only revoking the master key, I would propose that we also revoke all subkeys and all user ids. That's also the way it is done in Enigmail and it could prevent some bad handling in other PGP implementations, e.g, if they only check for the validity of the subkey used for signing and ignore if the master key is revoked.

dschuermann commented:
We should also discuss about setting a revocation reason here with PGPSignatureSubpacketGenerator.setRevocationReason and RevocationReasonTags.
I would propose we always use RevocationReasonTags.NO_REASON

Examples:
http://t124106.encryption-bouncy-castle-development-csharp.encryptiontalk.info/openpgp-key-revocation-generations-t124106.html
and especially: https://github.com/darkoperator/Posh-OpenPGP/blob/master/Source/PGPHelper/PGPHelper/Revocation.cs

BTW: Regarding the method PGPSignatureSubpacketGenerator.setRevocationKey: @Valodim isn't this the delegated revocation key? I haven't looked into it in detail...

dschuermann commented:
One last thing:
We should handle revocation of YubiKey keys. Currently OK crashes after entering the PIN:
```
Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_REVOKE, mParameters=[cd5b c031 bda0 d889], mIndent=0}
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/-3649111753186289527/keys, proj=[has_secret])
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/-3649111753186289527/unified, proj=[fingerprint])
  2097             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_ED, mParameters=null, mIndent=0}
  2097             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_ED_FETCHING, mParameters=[0xcd5bc031bda0d889], mIndent=1}
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/-3649111753186289527/unified, proj=[has_any_secret, verified, privkey_data])
  2097             Keychain D  V  log: LogEntryParcel{mLevel=START, mType=MSG_MF, mParameters=[0xcd5bc031bda0d889], mIndent=0}
  2097             Keychain D  D  Send message by setProgress with progress=10, max=100
  2097             Keychain D  D  Send message by setProgress with progress=10, max=100
  2097             Keychain D  V  log: LogEntryParcel{mLevel=DEBUG, mType=MSG_MF_DIVERT, mParameters=null, mIndent=1}
  2097             Keychain D  D  Send message by setProgress with progress=37, max=100
  2097             Keychain D  V  log: LogEntryParcel{mLevel=INFO, mType=MSG_MF_SUBKEY_REVOKE, mParameters=[0xcd5bc031bda0d889], mIndent=1}
  2097             Keychain D  D  Send message by setProgress with progress=60, max=100
  2097             Keychain D  V  log: LogEntryParcel{mLevel=OK, mType=MSG_MF_REQUIRE_DIVERT, mParameters=null, mIndent=1}
  741           WindowManager  V  not Base app: Adding window Window{1cd8a54b u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.DeleteKeyDialogActivity} at 25 of 32
  2097             Keychain D  D  Handling result in OperationHelper success: false
  2097               Timeline  I  Timeline: Activity_launch_request id:org.sufficientlysecure.keychain.debug time:547287522
  741         ActivityManager  I  START u0 {cmp=org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.NfcOperationActivity (has extras)} from uid 10167 on display 0
  741           WindowManager  V  addAppToken: AppWindowToken{1b3319e6 token=Token{2b9d0041 ActivityRecord{2d9d6b28 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.NfcOperationActivity t2752}}} to stack=1 task=2752 at 2
  2097             Keychain D  D  NfcOperationActivity.onCreate
  2097             Keychain D  D  PassphraseCacheService.getCachedPassphrase() for masterKeyId -3649111753186289527, subKeyId -3649111753186289527
  1981             Keychain D  D  PassphraseCacheService.onStartCommand()
  1981             Keychain D  D  PassphraseCacheService.getCachedPassphraseImpl() for masterKeyId -3649111753186289527, subKeyId -3649111753186289527
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/-3649111753186289527/keys, proj=[has_secret])
  1981             Keychain D  D  PassphraseCacheService: NOT using default YubiKey PIN
  1981             Keychain D  D  PassphraseCacheService: masterkey passphrase not (yet) cached, returning null
  1981             Keychain D  D  PassphraseCacheService: No passphrases remaining in memory, stopping service!
  2097               Timeline  I  Timeline: Activity_launch_request id:org.sufficientlysecure.keychain.debug time:547287584
  741         ActivityManager  I  START u0 {cmp=org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity (has extras)} from uid 10167 on display 0
  2097             Keychain D  D  BaseNfcActivity.onResume
  741           WindowManager  V  addAppToken: AppWindowToken{e785f7d token=Token{28f211d4 ActivityRecord{148a4b27 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity t2752}}} to stack=1 task=2752 at 3
  2097             Keychain D  D  NfcForegroundDispatch has been enabled!
  2097             Keychain D  D  BaseNfcActivity.onPause
  2097             Keychain D  D  NfcForegroundDispatch has been disabled!
  741           WindowManager  V  Adding window Window{13c672c3 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.NfcOperationActivity} at 25 of 32 (after Window{3a18e52f u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.DeleteKeyDialogActivity EXITING})
  283          SurfaceFlinger  W  couldn't log to binary event log: overflow.
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/-3649111753186289527/unified, proj=[has_any_secret, verified, privkey_data])
  2097             Keychain D  V  query(uri=content://org.sufficientlysecure.keychain.debug.provider/key_rings/find/subkey/-3649111753186289527, proj=[has_any_secret, verified, privkey_data])
  741           WindowManager  V  Adding window Window{36b07679 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity} at 25 of 32 (after Window{13c672c3 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.NfcOperationActivity})
  1413               LatinIME  I  Starting input. Cursor position = 0,0
  741         ActivityManager  I  Displayed org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity: +170ms (total +231ms)
  2097               Timeline  I  Timeline: Activity_idle id: android.os.BinderProxy@12b7ba34 time:547287787
  2097               Timeline  I  Timeline: Activity_idle id: android.os.BinderProxy@185ba321 time:547287788
  741                Timeline  I  Timeline: Activity_windows_visible id: ActivityRecord{148a4b27 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity t2752} time:547288283
  741                Timeline  I  Timeline: Activity_windows_visible id: ActivityRecord{2d9d6b28 u0 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.NfcOperationActivity t2752} time:547288287
  2097             Keychain D  D  Everything okay!
  2097         AndroidRuntime  D  Shutting down VM
  2097         AndroidRuntime  E  FATAL EXCEPTION: main
  2097         AndroidRuntime  E  Process: org.sufficientlysecure.keychain.debug, PID: 2097
  2097         AndroidRuntime  E  java.lang.NullPointerException: Attempt to read from field 'boolean org.sufficientlysecure.keychain.service.input.CryptoInputParcel.mCachePassphrase' on a null object reference
  2097         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:406)
  2097         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:357)
  2097         AndroidRuntime  E  at android.os.AsyncTask.finish(AsyncTask.java:636)
  2097         AndroidRuntime  E  at android.os.AsyncTask.access$500(AsyncTask.java:177)
  2097         AndroidRuntime  E  at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:653)
  2097         AndroidRuntime  E  at android.os.Handler.dispatchMessage(Handler.java:102)
  2097         AndroidRuntime  E  at android.os.Looper.loop(Looper.java:135)
  2097         AndroidRuntime  E  at android.app.ActivityThread.main(ActivityThread.java:5293)
  2097         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Native Method)
  2097         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Method.java:372)
  2097         AndroidRuntime  E  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
  2097         AndroidRuntime  E  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
  741         ActivityManager  W  Force finishing activity 1 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity
```

If you can not figure this out @adithyaphilip because you don't have a YubiKey, that's ok. We can then work on this together.

dschuermann commented:
The introduction of mPassphrase inside the crypto input parcel breaks decryption from k-9 mail:
```
 19474         AndroidRuntime  D  Shutting down VM
 19474         AndroidRuntime  E  FATAL EXCEPTION: main
 19474         AndroidRuntime  E  Process: org.sufficientlysecure.keychain.debug, PID: 19474
 19474         AndroidRuntime  E  java.lang.NullPointerException: Attempt to read from field 'boolean org.sufficientlysecure.keychain.service.input.CryptoInputParcel.mCachePassphrase' on a null object reference
 19474         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:406)
 19474         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:357)
 19474         AndroidRuntime  E  at android.os.AsyncTask.finish(AsyncTask.java:636)
 19474         AndroidRuntime  E  at android.os.AsyncTask.access$500(AsyncTask.java:177)
 19474         AndroidRuntime  E  at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:653)
 19474         AndroidRuntime  E  at android.os.Handler.dispatchMessage(Handler.java:102)
 19474         AndroidRuntime  E  at android.os.Looper.loop(Looper.java:135)
 19474         AndroidRuntime  E  at android.app.ActivityThread.main(ActivityThread.java:5293)
 19474         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Native Method)
 19474         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Method.java:372)
 19474         AndroidRuntime  E  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
 19474         AndroidRuntime  E  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
  741         ActivityManager  W  Force finishing activity 1 org.sufficientlysecure.keychain.debug/org.sufficientlysecure.keychain.ui.PassphraseDialogActivity
```

adithyaphilip commented:
* Fixed upload operation ignoring response code in https://github.com/open-keychain/open-keychain/pull/1438
* Revokes all subkeys and userIds with https://github.com/open-keychain/open-keychain/commit/4ed83bf301db389c0bbecc0c9200cded0f987faf
* Reason for crashes with k-9 and YubiKey revoke was because I hadn't handled cases where a `CryptoInputParcel` was not passed to `PassphraseDialogActivity`. https://github.com/open-keychain/open-keychain/commit/14c1e426210e6ff5faa8c222e01615603d618205 temporarily fixes the issue - just need to prevent caching of Yubikey PINs too.

TODO:
* Prevent caching of Yubikey PINs - just have to modify `NfcOperationActivity` to fit in with the rest of the `CryptoInputParcel` .
* Setting revocation reason.

adithyaphilip commented:
added revocation tags to revocations via PgpKeyOperation and squashed a few commits. We could ask the user what his reason for revocation was by providing a spinner with the possibilities in a dialog, after he presses the "Revoke and Upload" button, or is the reason something that isn't checked a lot?

dschuermann commented:
My opinion is that we shouldn't allow the user to select a reason here. It makes it overly complicated. The problem here is that RFC4880 really differentiates:
```
   If a key has been revoked because of a compromise, all signatures
   created by that key are suspect.  However, if it was merely
   superseded or retired, old signatures are still valid.  If the
   revoked signature is the self-signature for certifying a User ID, a
   revocation denotes that that user name is no longer in use.  Such a
   revocation SHOULD include a 0x20 code.
```

Again my opinion: Users often cannot decide properly if a key has been compromised or not. I myself would have a difficult time detecting a compromise, but if there is something suspect I would go with the worst case and revoke the key with tag 0x20 (compromised).
So, because other implementations handle the revocation reason, I re-decided and we should go with  KEY_COMPROMISED for master key and all subkeys and NO_REASON for all user ids.
@Valodim What's your opinion here?

dschuermann commented:
As discussed on IRC: Test how gnupg handles the signatures from revoked keys and then decide on our revocation reason.
Please also document in our code why we chose the reason.

dschuermann commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/NfcOperationActivity.java:
```diff
@@ -67,6 +68,14 @@ protected void onCreate(Bundle savedInstanceState) {
+        if (mInputParcel == null) {
+            // for compatibility when used from OpenPgpService
+            // (or any place other than CryptoOperationHelper)
+            mInputParcel = new CryptoInputParcel(mRequiredInput.mSignatureTime);
```
This looks like a change that could introduce future bugs. You assume that the NFC operation is a signing operation here.

adithyaphilip replied:
The current implementation of `CryptoOperationHelper` always assigns a date to `mSignatureTime`. If it isn't a signing operation and `mRequiredInput.mSignatureTime` is null, the current date is used to instantiate. Should we change this behavior of `CryptoOperationHelper` and introduce a `null` value for `mSignatureTime` when it isn't a signing operation? As of now it's declared as `final`, we'd just have to make it add-able like the rest of the attributes.

dschuermann replied:
Yes do as you proposed, sounds reasonable.

dschuermann commented:
@Valodim Please also review :)

adithyaphilip commented:
Looks like gnupg doesn't differentiate at all, just mentions the reason:
```
FOR compromised
gpg: WARNING: This key has been revoked by its owner!
gpg:          This could mean that the signature is forged.
gpg: reason for revocation: Key has been compromised
FOR nolonger
gpg: WARNING: This key has been revoked by its owner!
gpg:          This could mean that the signature is forged.
gpg: reason for revocation: Key is no longer used
FOR noreason
gpg: WARNING: This key has been revoked by its owner!
gpg:          This could mean that the signature is forged.
gpg: reason for revocation: No reason specified
FOR superseded
gpg: WARNING: This key has been revoked by its owner!
gpg:          This could mean that the signature is forged.
gpg: reason for revocation: Key is superseded
```
If there's a revocation comment the comment is just tacked on to the end. But there is no extra warning no matter which revocation tag.

dschuermann commented:
@adithyaphilip Great test! Then let's use NO_REASON. And please write a comment in the code about your findings how gnupg handles this (gnupg is always the reference implementation ;) )

Valodim commented:
just for the record, this is also what I gathered from the sourcecode

dschuermann commented:
@adithyaphilip Travis failure is my fault. Simply change the import as I did here: https://github.com/open-keychain/open-keychain/commit/56d599dd7648c4eafd87f8802f10cc86bcd67a4a

adithyaphilip commented:
@dschuermann @Valodim This is with respect to the change in `CryptoInputParcel` to make `mSignatureTime` optional. As of now the corrections that would be necessary:
* The default constructor `CryptoInputParcel` should say`mSignatureTime = null`.
* Parcelling/unparcelling should not require `mSignatureTime`. I thought we could either write a private `signaturePresent` boolean which if false would mean the time shouldn't be read, or write `mSignatureTime` directly which is `Serializable`. Former would be better, I suppose?
* A lot of the calls for signing operations used the default constructor since it created a date by itself, including `cryptoOperation()` in `CryptoOperationHelper`. This would require changing all those to `cryptoInputPrcel(Date signatureTime)` (and those that use `cryptoOperation()` should use `cryptoOperation(new CryptoInputParcel(Date signatureTime))`). Alternatively, instead of making the calls to the operations supply the `signatureTime`, we could make the operations themselves set the signatureTime if they require it. The advantage would be that there would be fewer changes to existing code. Something like:
```
if (cryptoInput.mSignatureTime == null) {
    cryptoInput.mSignatureTime = new Date();
}
```
Is this an acceptable way to go about it?

dschuermann commented:
* yes
* yes, use a boolean to check if signature time is present
* yes, use the constructor ``cryptoInputPrcel(Date signatureTime)`` for signing operations for clarity what is actually used here.

@Valodim will hopefully also comment.

Valodim commented:
will do, when time permits

adithyaphilip commented:
I've made all encrypt/sign, certify, revocation and edit operations require date with https://github.com/open-keychain/open-keychain/commit/771d65476647c4df0d88e95f8df80b201eeb9bf3. I guess it isn't necessary for subkey stripping or moving keys to card?

dschuermann commented:
If I sign a message with a Yubikey it still crashes with
```
 20737             Keychain D  D  Everything okay!
 20737         AndroidRuntime  D  Shutting down VM
 20737         AndroidRuntime  E  FATAL EXCEPTION: main
 20737         AndroidRuntime  E  Process: org.sufficientlysecure.keychain.debug, PID: 20737
 20737         AndroidRuntime  E  java.lang.NullPointerException: Attempt to read from field 'boolean org.sufficientlysecure.keychain.service.input.CryptoInputParcel.mCachePassphrase' on a null object reference
 20737         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:406)
 20737         AndroidRuntime  E  at org.sufficientlysecure.keychain.ui.PassphraseDialogActivity$PassphraseDialogFragment$4$1.onPostExecute(PassphraseDialogActivity.java:357)
 20737         AndroidRuntime  E  at android.os.AsyncTask.finish(AsyncTask.java:636)
 20737         AndroidRuntime  E  at android.os.AsyncTask.access$500(AsyncTask.java:177)
 20737         AndroidRuntime  E  at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:653)
 20737         AndroidRuntime  E  at android.os.Handler.dispatchMessage(Handler.java:102)
 20737         AndroidRuntime  E  at android.os.Looper.loop(Looper.java:135)
 20737         AndroidRuntime  E  at android.app.ActivityThread.main(ActivityThread.java:5293)
 20737         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Native Method)
 20737         AndroidRuntime  E  at java.lang.reflect.Method.invoke(Method.java:372)
 20737         AndroidRuntime  E  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
 20737         AndroidRuntime  E  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
```

dschuermann commented:
I am no longer able to delete two selected **public** keys. It shows "Nothing to delete" and in the log: "Secret keys can only be deleted individually".

dschuermann commented:
I also found this bug: https://github.com/open-keychain/open-keychain/issues/1447
But I think it's just a coincident that I found it using your pr.

dschuermann commented:
Looks like everything else works for me, so just the bug with the public key deletion should be fixed before merge.

dschuermann commented:
Ah and your question: Yes, stripping or moving keys to card should not require the sig time.

adithyaphilip commented:
Fixed the public key multi-delete bug and moved secret key multi-delete error logic into `DeleteKeyDialogActivity`.

dschuermann commented:
Nice work!

adithyaphilip commented:
Thank you :)

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/results/OperationResult.java:
```diff
@@ -760,6 +762,13 @@ public void onAction() {
+        MSG_REVOKE_ERROR_MULTI_SECRET (LogLevel.DEBUG, R.string.msg_revoke_error_multi_secret),
+        MSG_REVOKE_ERROR_NOT_FOUND (LogLevel.DEBUG, R.string.msg_revoke_error_multi_secret),
+        MSG_REVOKE (LogLevel.DEBUG, R.string.msg_revoke_key),
+        MSG_REVOKE_KEY_FAIL (LogLevel.ERROR, R.string.msg_revoke_key_fail),
```
ERROR_

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/results/OperationResult.java:
```diff
@@ -760,6 +762,13 @@ public void onAction() {
 
+        MSG_REVOKE_ERROR_EMPTY (LogLevel.ERROR, R.string.msg_revoke_error_empty),
+        MSG_REVOKE_ERROR_MULTI_SECRET (LogLevel.DEBUG, R.string.msg_revoke_error_multi_secret),
+        MSG_REVOKE_ERROR_NOT_FOUND (LogLevel.DEBUG, R.string.msg_revoke_error_multi_secret),
```
not ERROR_

Valodim commented:
DeleteKeyDialogActivity has unhandled warnings
